core/src/main/scala/cats/syntax/either.scala

```diff
@@ -160,6 +160,15 @@ final class EitherOps[A, B](private val eab: Either[A, B]) extends AnyVal {
       case Right(b) => Validated.valid(b)
     }
 
+  /**
+   * Generic implementation of [[toValidatedNel]] for any F[_] with provided Applicative instance. 
+   */
+  def toValidatedA[F[_]: Applicative, AA >: A]: Validated[F[AA], B] =
```

Comment:
I wonder if we are being too literal here. The real goal is to have a `Semigroup` on `F[AA]` isn't it?

So, I kind of wonder if what is going on is actually an `Alternative[F]` so that we can do both `pure(a)` and `combineK`.

What do you think about that? It is more restrictive, but does steer towards what is going on with Validated.

Reply:
@johnynek I'm not sure I'm catching your point here. I thought that the goal is to provide a handy conversion
```
Either[A, B] => Validated[F[A], B]
```
for some `F[_]`. but I'm not sure – what can we do with `combineK` in here?

Note, that `Alternative` is not implemented for non-empty collections like `NonEmptyList`, thus `NonEmptyAlternative` should be used instead.

Reply:
On the other hand, I have some usability concerns for such an implementation of `toValidatedA` in `syntax`. Consider that we would like to transform `Either[String, MyCustomModel]` into `Validated[NonEmptyVector[String], MyCustomModel]`. So we have to do the following for that:
```scala
val eee: Either[String, MyCustomModel] = ???

eee.toValidatedA[NonEmptyVector, MyCustomModel]
```
In other words, we have to specify not only the `F[_]` type but also the `Either`'s "right" type there. Even if it is supposed to be the same (and in most cases, it will). So I would say, it is not that convenient. I would rather prefer to write just
```scala
eee.toValidatedA[NonEmptyVector]
```
instead.

Reply:
I wanted to use `Monoid`/`MonoidK` first, but then I instantly realized all `NoneEmpty...` collections won't work :c 

Regarding usability, I completely agree with @satorg. I copied the signature from `def toValidatedNel[AA >: A]: ValidatedNel[AA, B]` where `[AA >: A]` seems redundant (correct me if I'm wrong). 
Therefore I have changed the `def toValidatedA` to 
```
  def toValidatedA[F[_]: Applicative]: Validated[F[A], B] =
    eab match {
      case Left(a)  => Validated.invalidA(a)
      case Right(b) => Validated.valid(b)
    }
```
Does it seem alright to you, guys?

Reply:
I think, it still should be feasible to have both `[AA >: A]` parametrization and good user experience for `.toValidatedA[F]`. However, more sophisticated syntax might be necessary for that, something like:
<sub>(disclaimer: not really tested)</sub>
```scala
final class ToValidatedApply[F[_], A, B](eab: Either[A, B]) extends AnyVal {
  def apply[AA >: A](implicit F: Applicative[F]) = eab match {
      case Left(a)  => Validated.invalidA(a)
      case Right(b) => Validated.valid(b)
  }
}

...

def toValidatedA[F[_]]: ToValidatedApply[F, A, B] = new ToValidatedApply[F, A, B](eab)
```
But I am not 100% sure if it is really worth it.

Reply:
I feel like from the user experience it's more convenient to do something like `either.leftMap(..).toValidatedA` or `either.leftWiden[AA]` than calling `ToValidatedApply#apply` explicitly.

Reply:
> @johnynek I'm not sure I'm catching your point here. I thought that the goal is to provide a handy conversion
> 
> ```
> Either[A, B] => Validated[F[A], B]
> ```
> 
> for some `F[_]`. but I'm not sure – what can we do with `combineK` in here?
> 
> Note, that `Alternative` is not implemented for non-empty collections like `NonEmptyList`, thus `NonEmptyAlternative` should be used instead.

My point is that we shouldn't forget what `Validated` is for. It is 100% so that there is an `Applicative` that combines the left values. It isn't just another datastructure. It is isomorphic to `Either`. If we had `opaque type` when we started cats, we would have used that and have `opaque type Validate[+A, +B] = Either[A, B]`. It is purely a mechanism to select which Applicative to use.

So, given that the use of`Validated[A, B]` relies on `Semigroup[A]` and that we are talking about functions which do: `Either[A, B] => Validated[F[A], B]`, then we are presuming `Semigroup[F[A]]` for all A. But that's the same thing as `SemigroupK[F]`. And since we need `pure` to do the `A => F[A]`, what it seems like we really need is `NonEmptyAlternative[F]` (as you corrected me Alternative doesn't work because we don't want to require `def empty[A]: F[A]`.

Anyway... the hole in this argument is that there is nothing that promises that the implicit `Semigroup[F[A]]` will be the same as `NonEmptyAlternative[F].algebra[A]` which is what we would really want to tie this all together.

In that case, I wouldn't actually make the change I am suggesting, but I do think it's a bit ugly that we don't have any laws for these functions except that they should behave as we say (as opposed having some rule about how `invalidA(x).product(invalidA(y))` should behave...

For instance, I bet we have:
```scala
invalidA(x).product(invalidA(y)) == invalid(pure(x).combineK(pure(y)))
```
in practice...